### PR TITLE
Refactor estimate_tune_results

### PR DESCRIPTION
**Pull request by topepo**
The existing `estimate_tune_results()` was used to extract out the average performance metrics. 

This PR expands its scope a little to allow it to do the same for the `.extracts` column. Getting that data in the right format is a lot more complex and should only be used in well-controlled circumstances (such as those planned for some new features). This does not supersede the `collect_extracts()` function; the new feature is for specific future usage. 

Also 

 * exports a developer function to get the mapping between tuning parameters, `.config`, and (potentially) `.iter`. 
 * Updates another error message for failed model with the "Run `show_notes(.Last.tune.result)` for more information." verbiage

### CODE REVIEW

File: R/collect.R [hunk 1]
```diff
@@ -368,51 +368,75 @@ collector <- function(x, coll_col = ".predictions") {
 #' @export
 #' @keywords internal
 #' @rdname empty_ellipses
-estimate_tune_results <- function(x, ...) {
+# Get relationship between the parameter values, .config, and (potentially) .iter
+config_key_from_metrics <- function(x) {
```
Comment (simonpcouch) [inline R/collect.R:372]: Not sure what the rules are for when internal-ish functions get a period prefix—is this one of those cases?

File: R/collect.R [hunk 2]
```diff
@@ -368,51 +368,75 @@ collector <- function(x, coll_col = ".predictions") {
...
 
   all_bad <- is_cataclysmic(x)
   if (all_bad) {
-    rlang::abort("All of the models failed. See the .notes column.")
+    rlang::warn("All models failed. Run `show_notes(.Last.tune.result)` for more information.")
   }
 
-  tibble_metrics <- purrr::map_lgl(x$.metrics, tibble::is_tibble)
-  x <- x[tibble_metrics, ]
+  # The the mapping of tuning parameters and .config.
```
Comment (simonpcouch) [inline R/collect.R:394]:
```suggestion
  # The mapping of tuning parameters and .config.
```

**Review by simonpcouch — COMMENTED** [review]
A few notes on smaller changes that this PR makes, but I'm in favor of the main idea! Glad that we're de-duplicating machinery to grab that `param_names` object.

Since this doesn't introduce any calls that tap into the `if (col_name == ".extracts") {}` body, I'm assuming that functionality will be tested where it's applied.

File: R/collect.R [hunk 3]
```diff
@@ -368,51 +368,75 @@ collector <- function(x, coll_col = ".predictions") {
...
-  if (".config" %in% param_names) {
-    arrange_names <- c(".iter", ".config")
-    arrange_names <- arrange_names[(arrange_names %in% param_names)]
-    join_names <- param_names[!(param_names %in% arrange_names)]
-    x <- dplyr::inner_join(
-      dplyr::select(x, !dplyr::all_of(arrange_names)),
-      x,
-      by = c(join_names, ".metric", ".estimator", "mean", "n", "std_err")
-    ) %>%
-      dplyr::arrange(!!!rlang::syms(arrange_names))
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/collect.R:414]: I'll fix that. Thanks

File: R/tune_bayes.R [hunk 4]
```diff
@@ -293,8 +290,13 @@ tune_bayes_workflow <-
       return(out)
     })
 
+    # Get the averaged resampling stats before stripping attributes
     mean_stats <- estimate_tune_results(unsummarized)
 
+    # Strip off `tune_results` class and drop all attributes since
+    # we add on an `iteration_results` class later.
+    unsummarized <- new_bare_tibble(unsummarized)
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/tune_bayes.R:298]: Yes.

File: R/collect.R (hunk 1 above)
Comment (topepo) [reply R/collect.R:372]: Good point. I'll add a .

File: R/tune_bayes.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonpcouch) [reply R/tune_bayes.R:298]: Okay, looks good.👍

File: R/collect.R [hunk 5]
```diff
@@ -368,51 +368,75 @@ collector <- function(x, coll_col = ".predictions") {
...
+#' @keywords internal
+#' @rdname empty_ellipses
+estimate_tune_results <- function(x, col_name = ".metrics", ...) {
+  param_names <- .get_tune_parameter_names(x)
+  id_names <- grep("^id", names(x), value = TRUE)
 
   all_bad <- is_cataclysmic(x)
   if (all_bad) {
-    rlang::abort("All of the models failed. See the .notes column.")
+    rlang::warn("All models failed. Run `show_notes(.Last.tune.result)` for more information.")
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/collect.R:391]: An oversight. Changing it to an error.

File: NEWS.md [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
+
 * Extended `show_best()`, `select_best()`, `select_by_one_std_error()`, `select_by_pct_loss()` to accommodate metrics with a target value of zero (notably, `yardstick::mpe()` and `yardstick::msd()`). (#243)
+
 * Clarified error messages in `select_by_*` functions. Error messages now only note entries in `...` that are likely candidates for failure to `arrange()`, and those error messages are no longer duplicated for each entry in `...`.
 
-* Improves condition handling for errors that occur during extraction from workflows. While messages and warnings were appropriately handled, errors occurring due to mis-specified `extract()` functions being supplied to `control_*()` functions were silently caught. As with warnings, errors are now surfaced both during execution and at `print()` (#575).
+* Improves condition handling for errors that occur during extraction from workflows. While messages and warnings were appropriately handled, errors occurring due to misspecified `extract()` functions being supplied to `control_*()` functions were silently caught. As with warnings, errors are now surfaced both during execution and at `print()` (#575).
```
Comment (simonpcouch) [inline NEWS.md:9]: Noting there's a duplicate change here in #586.

**Review by simonpcouch — APPROVED** [review]

**Pull request merged**